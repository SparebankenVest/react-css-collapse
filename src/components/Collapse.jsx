import React, { useRef } from 'react';
import { string, bool, func, shape } from 'prop-types';
import useCollapse from './useCollapse';

function Collapse({
  isOpen,
  onRest,
  style: initialStyle,
  transition,
  className,
  ...rest
}) {
  const content = useRef(null);
  const { setIsExpandedStyle, setIsCollapsedStyle, style } = useCollapse({
    isOpen,
    content,
  });

  const onTransitionEnd = (e) => {
    if (e.target === content.current && e.propertyName === 'height') {
      if (isOpen) {
        setIsExpandedStyle();
      } else {
        setIsCollapsedStyle();
      }
      if (onRest) {
        onRest();
      }
    }
  };

  const styles = {
    willChange: 'height',
    transition,
    ...initialStyle,
    ...style,
  };

  return (
    <div
      ref={content}
      style={styles}
      className={className}
      onTransitionEnd={onTransitionEnd}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
    />
  );
}

Collapse.defaultProps = {
  isOpen: false,
  onRest: null,
  style: null,
  className: 'react-css-collapse-transition',
  transition: 'height 250ms cubic-bezier(0.4, 0, 0.2, 1)',
};

Collapse.propTypes = {
  isOpen: bool,
  onRest: func,
  style: shape({}),
  className: string,
  transition: string,
};

export default Collapse;
