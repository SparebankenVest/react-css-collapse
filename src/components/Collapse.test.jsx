import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Collapse from './Collapse';

function transitionEndEventWithHeightProperty(el) {
  const event = new window.Event('transitionend', {
    bubbles: true,
    cancelable: true,
  });
  event.propertyName = 'height';
  fireEvent(el, event);
}

describe('Collapse', () => {
  beforeEach(() => {
    jest.spyOn(window, 'requestAnimationFrame').mockImplementation((cb) => {
      cb();
    });
  });

  afterEach(() => {
    window.requestAnimationFrame.mockRestore();
  });

  test('default', async () => {
    let component;
    await act(async () => {
      component = render(<Collapse data-testid="collapse" />);
    });
    const el = component.getByTestId('collapse');
    expect(el).toHaveStyle(
      'overflow: hidden; visibility: hidden; height: 0px;',
    );
    expect(el).toHaveClass('react-css-collapse-transition');
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with children', async () => {
    let component;
    await act(async () => {
      component = render(<Collapse data-testid="collapse">children</Collapse>);
    });
    const el = component.getByTestId('collapse');
    expect(el).toHaveTextContent('children');
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with class', async () => {
    let component;
    await act(async () => {
      component = render(
        <Collapse data-testid="collapse" className="className" />,
      );
    });
    const el = component.getByTestId('collapse');
    expect(el).toHaveClass('className');
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with style', async () => {
    let component;
    await act(async () => {
      component = render(
        <Collapse
          data-testid="collapse"
          style={{ transition: 'height 1337ms cubic-bezier(.4, 0, .2, 1)' }}
        />,
      );
    });
    const el = component.getByTestId('collapse');
    expect(el).toHaveStyle(
      'will-change: height; overflow: hidden; visibility: hidden; height: 0px; transition: height 1337ms cubic-bezier(.4, 0, .2, 1);',
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with transition prop', async () => {
    let component;
    await act(async () => {
      component = render(
        <Collapse
          data-testid="collapse"
          transition="height 1337ms cubic-bezier(.4, 0, .2, 1)"
        />,
      );
    });
    const el = component.getByTestId('collapse');
    expect(el).toHaveStyle(
      'will-change: height; overflow: hidden; visibility: hidden; height: 0px; transition: height 1337ms cubic-bezier(.4, 0, .2, 1);',
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('with aria and data props', async () => {
    let component;
    await act(async () => {
      component = render(
        <Collapse aria-hidden data-any-value="1337" data-testid="collapse" />,
      );
    });
    const el = component.getByTestId('collapse');
    expect(el).toHaveAttribute('data-any-value', '1337');
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('when open before transition end', async () => {
    let component;
    await act(async () => {
      component = render(<Collapse isOpen data-testid="collapse" />);
    });
    const el = component.getByTestId('collapse');
    expect(el).toHaveStyle(
      'overflow: visible; visibility: visible; height: auto;',
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('when open after transition end', async () => {
    let component;

    await act(async () => {
      component = render(<Collapse isOpen data-testid="collapse" />);
    });
    const el = component.getByTestId('collapse');
    transitionEndEventWithHeightProperty(el);
    expect(el).toHaveStyle(
      'overflow: visible; visibility: visible; height: auto;',
    );
    expect(component.asFragment()).toMatchSnapshot();
  });

  test('lifecycle', async () => {
    // Render default collapsed
    const { getByTestId, rerender } = render(
      <Collapse data-testid="collapse" />,
    );
    expect(getByTestId('collapse')).toHaveStyle(
      'visibility: hidden; overflow: hidden;',
    );

    // Expand
    rerender(<Collapse data-testid="collapse" isOpen />);
    const el = getByTestId('collapse');
    // Style during transition
    expect(el).toHaveStyle('visibility: visible; overflow: hidden;');
    // Finished expand transition
    transitionEndEventWithHeightProperty(el);
    // Style after transition
    expect(el).toHaveStyle(
      'visibility: visible; overflow: visible; height: auto;',
    );

    // Collapse
    rerender(<Collapse data-testid="collapse" />);
    const elCollapsed = getByTestId('collapse');
    // Style during transition
    expect(elCollapsed).toHaveStyle('visibility: visible; overflow: visible;');

    await act(async () => {
      // Collapse transition finished
      transitionEndEventWithHeightProperty(elCollapsed);
    });

    // Style after transition
    expect(elCollapsed).toHaveStyle(
      'visibility: hidden; overflow: hidden; height: 0px',
    );
  });
});
