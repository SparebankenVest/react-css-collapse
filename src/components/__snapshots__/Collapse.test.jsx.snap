// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse default 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    data-testid="collapse"
    style="will-change: height; transition: height 250ms cubic-bezier(0.4, 0, 0.2, 1); overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;

exports[`Collapse when open after transition end 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    data-testid="collapse"
    style="will-change: height; transition: height 250ms cubic-bezier(0.4, 0, 0.2, 1); overflow: visible; visibility: visible; height: auto;"
  />
</DocumentFragment>
`;

exports[`Collapse when open before transition end 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    data-testid="collapse"
    style="will-change: height; transition: height 250ms cubic-bezier(0.4, 0, 0.2, 1); overflow: visible; visibility: visible; height: auto;"
  />
</DocumentFragment>
`;

exports[`Collapse with aria and data props 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="react-css-collapse-transition"
    data-any-value="1337"
    data-testid="collapse"
    style="will-change: height; transition: height 250ms cubic-bezier(0.4, 0, 0.2, 1); overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;

exports[`Collapse with children 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    data-testid="collapse"
    style="will-change: height; transition: height 250ms cubic-bezier(0.4, 0, 0.2, 1); overflow: hidden; visibility: hidden; height: 0px;"
  >
    children
  </div>
</DocumentFragment>
`;

exports[`Collapse with class 1`] = `
<DocumentFragment>
  <div
    class="className"
    data-testid="collapse"
    style="will-change: height; transition: height 250ms cubic-bezier(0.4, 0, 0.2, 1); overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;

exports[`Collapse with style 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    data-testid="collapse"
    style="will-change: height; transition: height 1337ms cubic-bezier(.4, 0, .2, 1); overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;

exports[`Collapse with transition prop 1`] = `
<DocumentFragment>
  <div
    class="react-css-collapse-transition"
    data-testid="collapse"
    style="will-change: height; transition: height 1337ms cubic-bezier(.4, 0, .2, 1); overflow: hidden; visibility: hidden; height: 0px;"
  />
</DocumentFragment>
`;
